pydantic_core/core_schema.py

```diff
@@ -181,24 +181,26 @@ def simple_ser_schema(type: ExpectedSerializationTypes) -> SimpleSerSchema:
     return SimpleSerSchema(type=type)
 
 
-class GeneralPlainSerializerFunction(Protocol):  # pragma: no cover
-    def __call__(self, __input_value: Any, __info: SerializationInfo) -> Any:
-        ...
-
-
-class FieldPlainSerializerFunction(Protocol):  # pragma: no cover
-    def __call__(self, __model: Any, __input_value: Any, __info: FieldSerializationInfo) -> Any:
-        ...
-
+# (__input_value: Any) -> Any
+GeneralPlainNoInfoSerializerFunction = Callable[[Any], Any]
+# (__input_value: Any, __info: FieldSerializationInfo) -> Any
+GeneralPlainInfoSerializerFunction = Callable[[Any, SerializationInfo], Any]
+# (__model: Any, __input_value: Any) -> Any
+FieldPlainNoInfoSerializerFunction = Callable[[Any, Any], Any]
+# (__model: Any, __input_value: Any, __info: FieldSerializationInfo) -> Any
+FieldPlainInfoSerializerFunction = Callable[[Any, Any, FieldSerializationInfo], Any]
+SerializerFunction = Union[
+    GeneralPlainNoInfoSerializerFunction,
+    GeneralPlainInfoSerializerFunction,
+    FieldPlainNoInfoSerializerFunction,
+    FieldPlainInfoSerializerFunction,
+]
 
-class GeneralPlainSerializerFunctionSchema(TypedDict):
-    type: Literal['general']
-    function: GeneralPlainSerializerFunction
 
-
-class FieldPlainSerializerFunctionSchema(TypedDict):
-    type: Literal['field']
-    function: FieldPlainSerializerFunction
+class SerializerFunctionSchema(TypedDict):
+    type: Literal['general', 'field']
+    info_arg: bool
```

Comment:
It seems that by collapsing two schemas into one we are losing the type safety of matching up the function signatures with the discriminant in the actual types, but we are still enforcing it via the functions used to create these objects. Is that right?

Reply:
it's not fully enforced by the functions at the moment.

The function names get extremely long if we fully qualify the functions, e.g. `general_plain_serializer_function_ser_schema` would become `general_no_info_plain_serializer_function_ser_schema`.